### PR TITLE
New SendLocation scripts

### PR DESCRIPTION
**Pull request by GetmanetsIrina**
- MandatoryOnly
- latitudeDegrees bound values
- longitudeDegrees bound values

"GitHub issue link" ~ "https://github.com/smartdevicelink/sdl_core/issues/1740"

### CODE REVIEW

**Comment by GetmanetsIrina** [comment]
@dboltovskyi , @AByzhynar , @aderiabin , @Itileda , @AStasiuk , please review scripts.

File: test_scripts/API/Navigation/SendLocation/017_SendLocation_MandatoryOnly_Float.lua [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+-- 2) on getting Navigation.SendLocation ("SUCCESS") response from HMI, respond with (resultCode: SUCCESS, success:true)
+-- to mobile application.
+---------------------------------------------------------------------------------------------------
+
+--[[ Required Shared libraries ]]
+local runner = require('user_modules/script_runner')
+local commonSendLocation = require('test_scripts/API/Navigation/commonSendLocation')
+
+--[[ Local Variables ]]
+local request_params = {
```
Comment (AStasiuk) [inline test_scripts/API/Navigation/SendLocation/017_SendLocation_MandatoryOnly_Float.lua]: @GetmanetsIrina, suggested to use float data  with 7 decimal digits after point (like in the Google maps)

**Comment by AStasiuk** [comment]
@GetmanetsIrina, please add in the PR description "GitHub issue link" ~ "https://github.com/SmartDeviceLink/sdl_core/issues/1740"

File: test_scripts/API/Navigation/SendLocation/015_SendLocation_latitudeDegrees_bound_values.lua [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    EXPECT_HMICALL("Navigation.SendLocation", params)
+    :Do(function(_,data)
+        --hmi side: sending Navigation.SendLocation response
+        self.hmiConnection:SendResponse(data.id, data.method, "SUCCESS", {})
+    end)
+
+    self.mobileSession1:ExpectResponse(cid, { success = true, resultCode = "SUCCESS" })
+    :ValidIf (function(_,data)
+        if data.payload.info then
+            print("SDL sent redundant info parameter to mobile App ")
```
Comment (dboltovskyi) [inline test_scripts/API/Navigation/SendLocation/015_SendLocation_latitudeDegrees_bound_values.lua]: @GetmanetsIrina Please put error message just after return as 2nd parameter. E.g.:
```
return false , "Error"
```

File: test_scripts/API/Navigation/SendLocation/015_SendLocation_latitudeDegrees_bound_values.lua [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    EXPECT_HMICALL("Navigation.SendLocation", params)
+    :Do(function(_,data)
+        --hmi side: sending Navigation.SendLocation response
+        self.hmiConnection:SendResponse(data.id, data.method, "SUCCESS", {})
+    end)
+
+    self.mobileSession1:ExpectResponse(cid, { success = true, resultCode = "SUCCESS" })
+    :ValidIf (function(_,data)
+        if data.payload.info then
```
Comment (dboltovskyi) [inline test_scripts/API/Navigation/SendLocation/015_SendLocation_latitudeDegrees_bound_values.lua]: @GetmanetsIrina From my point of view check on info parameter is redundant. Since this verification is already implemented in script 001 (successful case)

File: test_scripts/API/Navigation/SendLocation/015_SendLocation_latitudeDegrees_bound_values.lua [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+runner.Step("Activate App", commonSendLocation.activateApp)
+
+--[[ Test ]]
+runner.Title("Test")
+runner.Step("SendLocation_latitudeDegrees lower bound", send_location, { request_params_with_lower_bound })
+runner.Step("SendLocation_latitudeDegrees upper bound", send_location, { request_params_with_upper_bound })
+
+--[[ Postconditions ]]
+runner.Title("Postconditions")
+runner.Step("Stop SDL", commonSendLocation.postconditions)
```
Comment (dboltovskyi) [inline test_scripts/API/Navigation/SendLocation/015_SendLocation_latitudeDegrees_bound_values.lua]: @GetmanetsIrina Please put a new line at the end of the script

File: test_scripts/API/Navigation/SendLocation/017_SendLocation_MandatoryOnly_Float.lua (hunk 1 above)
Comment (dboltovskyi) [reply test_scripts/API/Navigation/SendLocation/017_SendLocation_MandatoryOnly_Float.lua]: :+1:

File: test_scripts/API/Navigation/SendLocation/017_SendLocation_MandatoryOnly_Float.lua [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+runner.Step("RAI, PTU", commonSendLocation.registerApplicationWithPTU)
+runner.Step("Activate App", commonSendLocation.activateApp)
+
+--[[ Test ]]
+runner.Title("Test")
+runner.Step("SendLocation - mandatory only - float values", send_location, { request_params })
+
+--[[ Postconditions ]]
+runner.Title("Postconditions")
+runner.Step("Stop SDL", commonSendLocation.postconditions)
```
Comment (AByzhynar) [inline test_scripts/API/Navigation/SendLocation/017_SendLocation_MandatoryOnly_Float.lua]: @GetmanetsIrina Add empty line at the end of each script

File: test_scripts/API/Navigation/SendLocation/015_SendLocation_latitudeDegrees_bound_values.lua [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+-- Requirements: https://github.com/smartdevicelink/sdl_requirements/blob/master/detailed_docs/TRS/embedded_navi/SendLocation_TRS.md
+--
+-- Requirement summary:
+-- In case mobile application sends valid SendLocation request, SDL must:
+-- 1) transfer Navigation.SendLocation to HMI
+-- 2) on getting Navigation.SendLocation ("SUCCESS") response from HMI, respond with (resultCode: SUCCESS, success:true)
+-- to mobile application.
+--
+-- Description:
+-- App sends SendLocation request with latitudeDegrees bound values.
```
Comment (Itileda) [inline test_scripts/API/Navigation/SendLocation/015_SendLocation_latitudeDegrees_bound_values.lua]: @GetmanetsIrina , please take a look at script 009 in SendLocation folder:
Seems that boundary values are checked already.

Also, this is a common comment as for atomic scripts.
I suggest to do not cut boundary checks by parameters.
Atomic test assumes there is only one test scenario verified. 

It is a very good to have separate scripts even by parameters, but as for me this is an overhead.
Each script = independent test scenario, think - it's enough.
Loops are not forbidden to iterate through params for the same test.

File: test_scripts/API/Navigation/SendLocation/017_SendLocation_MandatoryOnly_Float.lua [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+--[[ Local Functions ]]
+local function send_location(params, self)
+    local cid = self.mobileSession1:SendRPC("SendLocation", params)
+
+    params.appID = commonSendLocation.getHMIAppId()
+
+    EXPECT_HMICALL("Navigation.SendLocation", params)
+    :Do(function(_,data)
+        --hmi side: sending Navigation.SendLocation response
```
Comment (Itileda) [inline test_scripts/API/Navigation/SendLocation/017_SendLocation_MandatoryOnly_Float.lua]: @GetmanetsIrina , suggest to remove obvious comment.

File: test_scripts/API/Navigation/SendLocation/017_SendLocation_MandatoryOnly_Float.lua [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+local commonSendLocation = require('test_scripts/API/Navigation/commonSendLocation')
+
+--[[ Local Variables ]]
+local request_params = {
+    longitudeDegrees = 1.1,
+    latitudeDegrees = 1.1
+}
+
+--[[ Local Functions ]]
+local function send_location(params, self)
```
Comment (Itileda) [inline test_scripts/API/Navigation/SendLocation/017_SendLocation_MandatoryOnly_Float.lua]: @GetmanetsIrina , suggest to use one style in code - camelCase.

File: test_scripts/API/Navigation/SendLocation/017_SendLocation_MandatoryOnly_Float.lua [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+    params.appID = commonSendLocation.getHMIAppId()
+
+    EXPECT_HMICALL("Navigation.SendLocation", params)
+    :Do(function(_,data)
+        --hmi side: sending Navigation.SendLocation response
+        self.hmiConnection:SendResponse(data.id, data.method, "SUCCESS", {})
+    end)
+
+    self.mobileSession1:ExpectResponse(cid, { success = true, resultCode = "SUCCESS" })
```
Comment (Itileda) [inline test_scripts/API/Navigation/SendLocation/017_SendLocation_MandatoryOnly_Float.lua]: @GetmanetsIrina , please remove redundant spaces:
self.mobileSession1:ExpectResponse(cid, { success = true, resultCode = "SUCCESS" })

self.mobileSession1:ExpectResponse(cid, {success = true, resultCode = "SUCCESS"})

File: test_scripts/API/Navigation/SendLocation/017_SendLocation_MandatoryOnly_Float.lua [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    :ValidIf (function(_,data)
+        if data.payload.info then
+            print("SDL sent redundant info parameter to mobile App ")
```
Comment (Itileda) [inline test_scripts/API/Navigation/SendLocation/017_SendLocation_MandatoryOnly_Float.lua]: @GetmanetsIrina , suggest to a little bit correct 
print("SDL sent redundant \"info\" parameter to mobile App")

File: test_scripts/API/Navigation/SendLocation/017_SendLocation_MandatoryOnly_Float.lua (hunk 1 above)
Comment (GetmanetsIrina) [reply test_scripts/API/Navigation/SendLocation/017_SendLocation_MandatoryOnly_Float.lua]: @AStasiuk Please see changes in 47644b6 commit

File: test_scripts/API/Navigation/SendLocation/015_SendLocation_latitudeDegrees_bound_values.lua (hunk 6 above)
Comment (GetmanetsIrina) [reply test_scripts/API/Navigation/SendLocation/015_SendLocation_latitudeDegrees_bound_values.lua]: @Itileda in 009 SendLocation script boundary values are checked in all scope of parameters.
In my opinion test is very common, in case of defect with some parameter script result will be not fully informative.
Agree with joining of tests with bound values of all parameters in one script.

**Comment by GetmanetsIrina** [comment]
@Itileda, @dboltovskyi , @AStasiuk please find changed according to review comments in commit e1e23ce .

File: test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+
+local deliveryModeEnum = {
+    "PROMPT",
+    "DESTINATION",
+    "QUEUE"
+}
+
+
+--[[ Local Functions ]]
+local function sendLocation(parameter, subParam, value, self)
```
Comment (dboltovskyi) [inline test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua]: @GetmanetsIrina Please split this function into a few ones in order to get rid of logical branches and to make script clear

**Comment by GetmanetsIrina** [comment]
@dboltovskyi Please check changes in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/1684/commits/b2bb7ee9afbc5a3281d67829f8479dee7259726c commit

File: test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+        { key, value.upper})
+end
+
+runner.Step("SendLocation_address_array_max_size", sendLocation,{ "addressLines", addressLinesArrayUpperBoundSize})
+runner.Step("SendLocation_address_array_max_size_max_value", sendLocation,
+    { "addressLines", addressLinesArrayUpperBoundSizeValue})
+
+for _, value in pairs(deliveryModeEnum) do
+    runner.Step("SendLocation_deliveryMode_enum_" .. tostring(value), sendLocation,
+        { "deliveryMode",_, value})
```
Comment (dboltovskyi) [inline test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua]: @GetmanetsIrina Please get rid of '_' as values for function's parameters. Use nil instead. 
In this particular case according to definition of sendLocation(parameter, value, self) function underscore have to be removed. Otherwise script fails to execute with aborted status

File: test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+    EXPECT_HMICALL("Navigation.SendLocation", request_params)
+    :Do(function(_,data)
+        self.hmiConnection:SendResponse(data.id, data.method, "SUCCESS", {})
+    end)
+
+    self.mobileSession1:ExpectResponse(cid, { success = true, resultCode = "SUCCESS" })
+end
+
+local function sendLocationTimeStamp(paramVal, boundValue, self)
+    local params = {
```
Comment (dboltovskyi) [inline test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua]: @GetmanetsIrina Default values for mandatory parameters used in every of 4 functions. Suggest to move this definition to 'Local Variables' sections. Please don't forget to create a copy of original table inside of the function since the values are changed through the functions

File: test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+    end
+
+    EXPECT_HMICALL("Navigation.SendLocation", params)
+    :Do(function(_,data)
+        self.hmiConnection:SendResponse(data.id, data.method, "SUCCESS", {})
+    end)
+    :ValidIf (function(_,data)
+        local ErrorMessage = ""
+        local ValidationStatus = true
+        for i,_ in pairs(DateTimeParams) do
```
Comment (dboltovskyi) [inline test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua]: @GetmanetsIrina Underscore as 2nd not used parameter can be omitted

File: test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua [hunk 15]
```diff
@@ -0,0 +1,280 @@
...
+            if data.params.timeStamp[DateTimeParams[i].name] and
```
Comment (dboltovskyi) [inline test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua]: @GetmanetsIrina ```DateTimeParams[i].name``` is not defined and will be nil

File: test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+
+local locationImage = {lower = "a", upper = string255Char}
+
+local deliveryModeEnum = {
+    "PROMPT",
+    "DESTINATION",
+    "QUEUE"
+}
+
+
```
[the remark this answers is not in the document]
Comment (GetmanetsIrina) [reply test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua]: Done

File: test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua (hunk 13 above)
Comment (GetmanetsIrina) [reply test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua]: Done

File: test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua (hunk 14 above)
Comment (GetmanetsIrina) [reply test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua]: Done

File: test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua (hunk 15 above)
Comment (GetmanetsIrina) [reply test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua]: Done

File: test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua (hunk 12 above)
Comment (GetmanetsIrina) [reply test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua]: Done

**Comment by GetmanetsIrina** [comment]
@dboltovskyi please review changes in b9271a10d56677a43f63076504ba04552b10ebdd

File: test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+local deliveryModeEnum = {
+    "PROMPT",
+    "DESTINATION",
+    "QUEUE"
+}
+
+--[[ Local Functions ]]
+local function sendLocation(parameter, value, self)
+    local paramters  = commonFunctions:cloneTable(request_params)
```
Comment (dboltovskyi) [inline test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua]: @GetmanetsIrina Please fix typo "paramters"

File: test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua [hunk 18]
```diff
@@ -0,0 +1,271 @@
...
+    paramters["locationImage"] = {value = value, imageType = "DYNAMIC"}
+
+    local cid = self.mobileSession1:SendRPC("SendLocation", paramters)
+
+    paramters.appID = commonSendLocation.getHMIAppId()
+    local deviceID = commonSendLocation.getDeviceMAC()
+    paramters.locationImage.value = commonSendLocation.getPathToSDL() .. "storage/"
+    .. commonSendLocation.getMobileAppId(1) .. "_" .. deviceID .. "/" ..tostring(value)
+
+
```
Comment (dboltovskyi) [inline test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua]: @GetmanetsIrina Please remove extra new line

File: test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua [hunk 19]
```diff
@@ -0,0 +1,271 @@
...
+    local cid = self.mobileSession1:SendRPC("SendLocation", paramters)
+
+    paramters.appID = commonSendLocation.getHMIAppId()
+
+    EXPECT_HMICALL("Navigation.SendLocation", paramters)
+    :Do(function(_,data)
+        self.hmiConnection:SendResponse(data.id, data.method, "SUCCESS", {})
+    end)
+    :ValidIf (function(_,data)
```
Comment (dboltovskyi) [inline test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua]: @GetmanetsIrina Please remove extra space

File: test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua (hunk 17 above)
Comment (GetmanetsIrina) [reply test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua]: Done in 4875635ea9ce242266c4f0d7bbaece1731a76faf

File: test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua (hunk 19 above)
Comment (GetmanetsIrina) [reply test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua]: Done in 4875635ea9ce242266c4f0d7bbaece1731a76faf

File: test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua (hunk 18 above)
Comment (GetmanetsIrina) [reply test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua]: Done in 4875635ea9ce242266c4f0d7bbaece1731a76faf

**Comment by GetmanetsIrina** [comment]
@dboltovskyi please review changes in 4875635ea9ce242266c4f0d7bbaece1731a76faf

**Comment by GetmanetsIrina** [comment]
@AStasiuk please review changes according to comments.

File: test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua (hunk 19 above)
Comment (dboltovskyi) [reply test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua]: @GetmanetsIrina Please also remove extra space in lines  187, 229, 249

Comment (GetmanetsIrina) [reply test_scripts/API/Navigation/SendLocation/020_SendLocation_boundary_by_parameters.lua]: @dboltovskyi done in d0f8f738ed85818574f90acc1831e910b41ec179
